File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
import heapq
def solution(book_time):
    answer = 1
    
    book = [(int(s[:2])*60+int(s[3:]),int(e[:2])*60+int(e[3:])) for s,e in book_time]
    book.sort()
    
    h = []
    
    for s,e in book:
        if not h:
            heapq.heappush(h,e)
            continue
        if h[0]<=s:
            heapq.heappop(h)
        else:
            answer+=1
        heapq.heappush(h,e+10)
    return answer

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;
class Solution {
    public int solution(String[][] book_time) {
        int answer = 0;
        List<int[]> times = new ArrayList<>();
        for(int i=0; i<book_time.length; i++){
            times.add(new int[]{revertTime(book_time[i][0]),revertTime(book_time[i][1])});
        }
        Collections.sort(times, (a,b)-> a[0]-b[0]);
    
        
        PriorityQueue<Integer> q = new PriorityQueue<>();
        for(int[] time : times){
            int start = time[0];
            int end = time[1];
            
            if(!q.isEmpty()&& q.peek()+10 <= start){
                q.poll();
            }
            q.offer(end);
        }
        
        answer = q.size();
        
        return answer;
    }
    public static int revertTime(String t){
        String[] time = t.split(":");
        return Integer.parseInt(time[0])*60 + Integer.parseInt(time[1]);
    }
}